tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/TensorScope.java

```diff
@@ -0,0 +1,536 @@
+/*
+  Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ ==============================================================================
+ */
+package org.tensorflow;
+
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+
+/**
+ * A scope that can be used to manage tensor resources.  Any tensors created between a scope's creation and calling
+ * {@code close()} that haven't been detached or attached to a different scope are guaranteed to be closed with the
+ * scope (even if they are created in a sub-scope).  Tensors may be manually closed earlier without issue.
+ * <p>
+ * Tensors are automatically tracked on creation.  A tensor can me manually added to a scope with {@link
+ * TensorScope#attach(Tensor)} or {@link Tensor#attachToCurrent()}.  A tensor may only have one scope: if it currently
+ * has a scope when {@code attach} is called, it is removed from its original scope.
+ * <p>
+ * {@link Tensor#detach()} detaches the tensor from it's scope, requiring the user to close it manually or attach it to
+ * another scope.
+ * <p>
+ * Scopes will be inherited at thread creation, but further scope creation on different threads will be independent,
+ * other than having the same parent. Closing a scope will close it's children regardless of which threads they are on.
+ */
+public class TensorScope implements AutoCloseable {
+
+  private static final InheritableThreadLocal<TensorScope> currentScope = new InheritableThreadLocal<>();
+
+  public static TensorScope currentScope() {
+    TensorScope scope = currentScope.get();
+
+    if (scope == null || !scope.closed) {
+      return scope;
+    }
+
+    // scope could be closed in another thread, in which case this thread's currentScope wouldn't be updated
+    while (scope != null && scope.closed) {
+      scope = scope.parent;
+    }
+    currentScope.set(scope);
+    return scope;
+  }
+
+  /**
+   * Runs {@code block}, then closes any tensors created during its execution.
+   * <p>To release tensors, use {@link #withCleanup(Consumer)} or one of the {@code produceWithCleanup} methods.
+   */
+  public static void withCleanup(Runnable block) {
```

Comment:
Is there any test or examples for the usage of these new `withCleanup` methods? I don't see any in the unit tests.

Reply:
I haven't added any yet (I need to do a docs update, too), but they would have issues since I want to be able to check if the scope and tensors are closed.  I could always store them outside the lambda, but if someone is looking a the tests for correct usage, it would be confusing.

I'll add tests to the produce versions to ensure the results are released, but I think the regular ones are covered well enough by the manual closing tests.